docs/design/2020-01-10-shuffle-executor.md

```diff
@@ -0,0 +1,253 @@
+# Shuffle Executor
+
+- Author(s):         [pingyu](https://github.com/pingyu)  (Ping Yu)
+- Last updated:  2020-01-11
+
+## Abstract
+This article will describe the design of Shuffle Executor in TiDB.
+
+## Background
+As described in [#12966](https://github.com/pingcap/tidb/issues/12966), the performance of Window operator in TiDB has much space to improve.
+
+So we designed a Shuffle Executor, which is proposed [here](https://github.com/pingcap/tidb/pull/14238#issuecomment-569880893) by [SunRunAway](https://github.com/SunRunAway) (Feng Liyuan) , using multi-thread hash grouping, to run each Window partition in parallel.
+
+Furthmore, the Shuffle Executor is designed to be a general purpose executor, which can also be used for other scenario, such as Parallel Sort-Merge Join, in the future.
+
+
+## Rationale
+[Window operator](https://dev.mysql.com/doc/refman/8.0/en/window-functions.html) separates data into independent groups by `PARTITION BY` clause, and the independency makes parallel executing possible.
+
+The rationale of Shuffle Exector is simple: It splits and shuffles data into partitions by columns in `PARTITION BY` clause, executes Window functions on partitions in a parallel manner, and finally merges the results.
+
+![shuffle-rationale](imgs/shuffle-rationale.png)
+
+## Implementation
+
+The implementation consists of 3 parts:
+1. Physical Planning.
+2. Building Executors.
+3. Executing.
+
+### Planning
+
+We find the very position and very moment for "shuffle" in Physical Planning procedure.
+
+The very position is on top of the operator which can be "shuffled", i.e. Window operator so far.
```

Comment:
IMHO, shuffle is used to redistribute data: it changes the data property from 1 distribution to another distribution. In a multi-thread execution model, the operator which can change the data distribution can be `group by`, `join`, etc.

In order to correctly handle the `shuffle` operator, we should consider the distribution property in the physical optimization phase.

Reply:
It means that we should do "shuffle" according to the physical property, e.g. hash on unsorted data, streaming on sorted data. Isn't it ?

Reply:
@zz-jason 
Comment addressed, PTAL~